### PR TITLE
🏗 CircleCI workspace optimizations

### PR DESCRIPTION
**Pull request by danielrozenberg**
* Rename the restored CircleCI workspace to avoid doubly-storing the same files
* Call maybe_gracefully_halt.sh via curl instead of hard-coding it into the config file (requires #33734 to be merged first)
* Use the merge commit info file in the workspace (requires #33734)

### CODE REVIEW

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:58]: From my understanding of this code, it will take effect when an additional layer is added to our CI matrix. Are there any plans to do this? (Or is there something else at play?)

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:56]: This step has now become complex enough that I think it's time to move it into its own shell script and invoke it here via `curl`, similar to `'Compute Merge Commit'`. The main reason is that every edit to this file will force all in-flight PRs across the repo to rebase their PR branches, while edits to the shell script will not.

**Review by rsimha — COMMENTED** [review]
A couple comments below.

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:58]: Yes, I said that in the PR description :)

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:56]: I'm worried that using curl-and-execute would make this _harder_ to update. A PR that changes the bash file can't test itself because it'll still fetch the bash file from the main branch. How does that work?

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml:58]: I read the description. I was asking about what the upcoming change is that requires a split in the workspace. The main concern being that adding a third (fourth?) stage to CI has the potential to reduce overall throughput for the organization due to increased VM use.

Comment (danielrozenberg) [reply .circleci/config.yml:58]: wdym by "stage"? Are you talking about more steps inside a job, or another "layer" of jobs inside the workflow graph?

If it's the former, steps are cheap - they're just a way to group commands and outputs, but they run on the same shell instance.

If it's the latter, then that's not the thing I'm planning on changing. The reason I wanted to split the restored and the new workspace for each job is so that devs working on pr-check/ files (i.e., wg-infra) don't have to worry about micro-managing the workspace between the jobs. Incoming files are always going to be in /tmp/restored-workspace, outgoing files go in /tmp/workspace - and you won't have to worry about leaving existing files there and causing a workspace merge conflict

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:56]: Alright, I see your point... Let's do this "properly" by splitting this into two PRs

I think the `mv workspace` command should stay its own hard-coded step (because it's extremely CircleCI specific and could be confusing to future readers if it's not explicitly there), but the rest of the graceful-halt can be moved to a curl'd file. Agree?

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml:58]: > If it's the latter, then that's not the thing I'm planning on changing. The reason I wanted to split the restored and the new workspace for each job is so that devs working on pr-check/ files (i.e., wg-infra) don't have to worry about micro-managing the workspace between the jobs. Incoming files are always going to be in /tmp/restored-workspace, outgoing files go in /tmp/workspace - and you won't have to worry about leaving existing files there and causing a workspace merge conflict

Yes, my concern was indeed the latter, and this explanation is exactly what I was looking for. Thanks for indulging me, carry on!

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml:56]: I still say we should move all steps into a shell script and change this command to a single `curl` invocation. The rationale for this is in the screenshot below:

image

All our CircleCI specific business logic lives in these shell files. It has two advantages:
- Makes `config.yml` a lot less compact and manageable
- Eliminates the need for wholesale rebasing of in-flight PRs when changes are restricted to the shell scripts
- Since these files all live in the `.circleci/` directory, it's already clear they shouldn't be used anywhere else.

WDYT? 😃

Comment (danielrozenberg) [reply .circleci/config.yml:56]: Tested this PR by curling from my own branch and the [CI passed](https://app.circleci.com/pipelines/github/ampproject/amphtml/6477/workflows/acf8d6ef-45d5-4057-8095-cf2ea7096e19), I'll now push a commit with the corrected URL which will cause this PR to fail until we merge #33734

Comment (danielrozenberg) [reply .circleci/config.yml:56]: Missed your comment above. PTAL at the config file, do you think it makes sense to add the `mv` command into `maybe_gracefully_halt.sh`? It seems like the wrong place and would be unexpected for the unaware reader

Comment (rsimha) [reply .circleci/config.yml:56]: Good question. How about combining the two steps, roughly like this?

```yml
  restore_workspace_and_maybe_halt:
    steps:
      - attach_workspace:
          at: /tmp
      - run:
          name: 'Restore Workspace and Maybe Halt'
          command: curl -sS https://raw.githubusercontent.com/ampproject/amphtml/main/.circleci/restore_workspace_and_maybe_halt.sh | bash
```

Comment (danielrozenberg) [reply .circleci/config.yml:56]: I feel pretty strongly that the rename should be its own thing :/

Comment (rsimha) [reply .circleci/config.yml:44]: No worries, I'm happy to defer to you on this one. I think you've already done plenty to address the top-level concern of minimizing changes to the root `config.yml`.

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:99]: Can this empty `run` block be deleted?

**Review by rsimha — APPROVED** [review]
Pre-emptive LGTM based on all the discussions in this PR and others. Thanks for patiently answering all my questions!

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml]: Why yes. Yes it can.

**Pull request merged**
